app/src/main/java/nz/org/cacophony/cacophonometerlite/RecordAndUpload.java

```diff
@@ -256,9 +256,11 @@ public static boolean makeRecording(Context context,  long recordTimeSeconds, bo
     // Try to prepare recording.
     try {
         mRecorder.setAudioSource(MediaRecorder.AudioSource.MIC);
-        mRecorder.setOutputFormat(MediaRecorder.OutputFormat.THREE_GPP);
+        mRecorder.setOutputFormat(MediaRecorder.OutputFormat.MPEG_4);
         mRecorder.setOutputFile(filePath);
-        mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.AMR_NB);
+        mRecorder.setAudioEncodingBitRate(256000);
+        mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.HE_AAC);
+        mRecorder.setAudioSamplingRate(44100);
```

Comment:
I'm pretty sure this isn't creating an MP3 file so the ".mp3" extension is misleading. I think this is creating a file using the MP4 container format with AAC encoded audio inside. The most appropriate extension for that is ".m4a".

I tried playing the example file you attached to the email in Firefox and it would only play if I changed the extension to .m4a. Chrome doesn't seem to want to play it at all. You might want to try some other combination (maybe one of the other AAC encoders will work with Chrome?). If our phones are up to API level 21 (Lollipop) then Vorbis encoding in a WEBM container might be a good choice. 

See also the comments here: https://stackoverflow.com/a/33054794/195383

Reply:
Will look into changing extension.

Our 'reference' phone is API 22, but I am currently trying to keep the code working on API 9 to allow old phones to work.  So do we want to just bite the bullet and say 21 is the minimum, or allow older phones to send lower/different quality recordings, and then how will the server deal with different formats?  Also, my colleague Jason raised the issue that these higher quality recordings will consume more battery power sending the files - but I'm struggling to keep the phones going on a solar panel as it is, so Grant is thinking we give up on solar panels for now and just assume mains....  Thoughts?

Reply:
We could have a few different preset quality options (low , med, high) that can be changed in the settings for the power concerns. I would be curious as to how much more power this actually uses.
For the minimum API we can have the older phones continue to use the AMR_NB and the newer phones using the other encoder that we decide to use.
Do you want to have a go changing the encoder to get a format that works on firefox and Chrome? As Menno said I think WEBM and Vorbis are probably good options for this.
Once we have a format that we want to use we can update the server so it won't process those files into an mp3.

Reply:
I've looked at a few things:

1) It seems odd that Chrome does not play the new file format with your code changes (changing extension does not help).  From what I've read, Chrome seems to support the most types.

2) In Android code, it shows (in context popup) that Vorbis is only supported in Android 6 (23) which is higher than 22 of our standard phone.   Menno, your suggestion to use Vorbis encoding in a WEBM container does not seem to be an option that I can see for 22 - but correct me if I'm wrong please.

3) I changed the AudioEncoder to .AAC_ELD, and the resulting file (using output format of MPEG_4 and .m4a extension) played in Chrome BUT in NOT Edge or Firefox!!!  The file was also four times the size (1 minute - 1891 kB  (.AAC_ELD), cf 481 kB (.HE_AAC) ) - both recording the same white noise.  I'll ask Jason to see if he can tell the difference in quality.

4)I'm thinking:
a) We use .HE_AAC and perhaps don't support Chrome just yet - do the Dilbert think and wait for the world to change :-)
b) Implement Cameron's suggestion of supporting a variety of qualities, including looking at using AMR-WB (9 rates from 6.60 kbit/s to 23.85 kbit/s sampled @ 16kHz) for Android 2 (API 5) and above as this is better than AMR-NB (4.75 to 12.2 kbps sampled @ 8kHz) which I think is what we have been using.  And when I have a phone running Android 6 look at Vorbis

Comments please!

Reply:
Weird that the IDE says that Vorbis+WEBM isn't supported until API level 23. According to the docs they were added in API level 21:
* https://developer.android.com/reference/android/media/MediaRecorder.OutputFormat#WEBM
* https://developer.android.com/reference/android/media/MediaRecorder.AudioEncoder.html#VORBIS

Is it possible to try it anyway?

With regards to Chrome, it appears there's a fair number of people reporting issues playing some M4A files. It seems like it could be software patent related.

Have you tried the straight `AAC` encoder (i.e. not `HE_ACC` or `AAC_ELD`)? Maybe it doesn't sound as good but it's worth a try.

Regarding supporting older phones as well as new ones, is there a way for the app to dynamically adjust the output type depending on the Android API level? Having 2 or 3 formats that we use depending on how modern the phone is would be workable. We can handle the differences on the server side, leaving the file alone where possible. 

I'm pretty sure the minimum we should use is AMR_WB (wideband) in a 3GP container instead of AMR_NB as we're currently using. Our experiments yesterday showed that this can sound *much* better than the current files, as long as a suitably high bit rate is selected. 3GPs will probably still have to be converted server side in order to be playable in browsers though. We should keep the originals though.

@timhot @CameronRP have either of you tried to play a 3GP file in the browser? It might just work.

Reply:
AMR_NB is designed for phone call quality recordings (i.e. preserving frequency ranges up to 8kHz).

Reply:
Just tried using WEBM and VORBIS but got an error: MediaRecorder : start failed: -38 - couldn't find what it meant.

 fileName += ".m4a";
....
 mRecorder.setOutputFormat(MediaRecorder.OutputFormat.WEBM); // WEBM added in API level 21
 mRecorder.setAudioEncodingBitRate(130000);
  mRecorder.setAudioEncoder(MediaRecorder.AudioEncoder.VORBIS); // VORBIS added in API 21
   mRecorder.setAudioSamplingRate(16000);